src/rpcclient/rpcclient/darwin/duet.py

```diff
@@ -0,0 +1,277 @@
+                    self._client.symbols.method_getName(methods[i])
+                ).peek_str()
+                stream = _DKSystemEventStreams.objc_call(method_name)
+                self.streams[stream.objc_call('name').py()] = stream
+
+    def query_events(self, stream: str) -> Optional[List['DKEvent']]:
+        """Retrieve all events for a single stream.
+
+        :param stream: Stream identifier.
+        :returns: List of events or ``None`` when the stream is empty.
+        """
+        raw_events = self._query_raw_events(stream)
+        event_count = raw_events.objc_call('count')
+        return (
+            [DKEvent(raw_events.objc_call('objectAtIndex:', i)) for i in range(event_count)]
+            if event_count != 0
+            else None
+        )
+
+    def query_events_streams(self, streams: Optional[List[str]] = None) -> dict[str, List['DKEvent']]:
+        """Query multiple streams at once.
+
+        :param streams: Specific streams to query. ``None`` queries all known streams.
+        :returns: Mapping of stream name to event list (absent when stream empty).
+        """
+        events: dict[str, List[DKEvent]] = {}
+        streams = list(self.streams.keys()) if streams is None else streams
+        for stream in streams:
+            events_for_stream = self.query_events(stream)
+            if events_for_stream is not None:
+                events[stream] = events_for_stream
+        return events
+
+    def _query_raw_events(self, stream: str) -> DarwinSymbol:
+        """Build and execute a ``_DKEventQuery`` limited to *stream*.
+
+        :param stream: Stream name.
+        :returns: ObjectiveC array containing raw ``_DKEvent`` objects.
+        """
+        query = self._client.symbols.objc_getClass('_DKEventQuery').objc_call('new')
+        query.objc_call('autorelease')
+        with self._client.safe_malloc(8) as buf:
+            buf[0] = self.streams[stream]
+            arr = self._client.symbols.objc_getClass('NSArray').objc_call(
+                'arrayWithObjects:count:', buf, 1
+            )
+            query.objc_call('setEventStreams:', arr)
+            return self._execute_query(query)
+
+    def _execute_query(self, query: DarwinSymbol) -> DarwinSymbol:
+        """Run *query* through the concrete :pyattr:`knowledge_store` implementation.
+
+        :param query: Configured ``_DKEventQuery`` instance.
+        :returns: ObjectiveC array with results.
+        :raises RpcClientException: If :pyattr:`knowledge_store` is not yet initialized.
+        """
+        if self.knowledge_store is None:
+            raise RpcClientException('KnowledgeStore not initialized')
+        with self._client.safe_malloc(8) as error:
+            return self.knowledge_store.objc_call('executeQuery:error:', query, error)
+
+    @abstractmethod
+    def __enter__(self) -> 'KnowledgeStoreContext':
+        """Enter the context manager.
+
+        :returns: Self reference.
+        """
```

Comment:
Though this is correct python syntax, I consider it less-readable then adding `pass`